### PR TITLE
fixed BOffcanvas and BModal hide event

### LINKED ISSUE #1526: BOffcanvas hide event when no needed
### Describe the bug

Even with ```no-close-on-esc``` or ```no-close-on-backdrop``` attributes BOffcanvas fires ```hide``` event when pressing ESC or clicking on backdrop

### Reproduction

https://stackblitz.com/edit/github-bzyfdw?file=src%2Fmain.js,src%2FApp.vue

### System Info

```Shell
latest bootstrap-vue-next
```

### Used Package Manager

pnpm

### Validations

- [X] Have tested with the latest version. This is still alpha version and sometime things change rapidly.
- [X] Follow our [Code of Conduct](https://github.com/bootstrap-vue-next/bootstrap-vue-next/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://github.com/bootstrap-vue-next/bootstrap-vue-next/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A, please open a GitHub Discussion instead.
- [X] The provided reproduction is a [minimal reproducible](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by makroxyz**
# Describe the PR

fixes #1526

### CODE REVIEW

**Comment by VividLemon** [comment]
```ts
const hide = (trigger = '') => {
  const event = buildTriggerableEvent('hide', {cancelable: trigger !== '', trigger})

  if (
    event.defaultPrevented ||
    (trigger === 'backdrop' && noCloseOnBackdropBoolean.value) ||
    (trigger === 'esc' && noCloseOnEscBoolean.value)
  ) {
    emit('hide-prevented', event)
    return
  }

  if (trigger === 'close') {
    emit(trigger, event)
  }
  if (trigger === 'esc') {
    emit(trigger, event)
  }
  emit('hide', event)

  if (event.defaultPrevented) {
    emit('hide-prevented', event)
    return
  }

  modelValue.value = false
}

const show = () => {
  const event = buildTriggerableEvent('show', {cancelable: true})
  emit('show', event)
  if (event.defaultPrevented) {
    modelValue.value = false
    emit('show-prevented', event)
    return
  }
  modelValue.value = true
}
```

Other files suffer from a similar issue.

**Comment by makroxyz** [comment]
Sorry I don't understand if this is expected or we need to fix other files too

**Comment by VividLemon** [comment]
> Sorry I don't understand if this is expected or we need to fix other files too

The codeblock I gave is likely the solution you should update to. In addition, multiple other files should be fixed with that fix as well.

**Comment by makroxyz** [comment]
Ok, I'm not on PC now so I thought was code from my PR I also "switched"
the if sequence.
Tomorrow I will check for other components

Il gio 19 ott 2023, 16:48 Issayah ***@***.***> ha scritto:

**Comment by makroxyz** [comment]
Hi I found also BModal  and fixed it.

**Comment by makroxyz** [comment]
@VividLemon 

I think this is better and modal.spec.ts doesn't fail
```js
const hide = (trigger = '') => {
  const event = buildTriggerableEvent('hide', {cancelable: trigger !== '', trigger})

  if (trigger === 'ok') {
    emit(trigger, event)
  }
  if (trigger === 'cancel') {
    emit(trigger, event)
  }
  if (trigger === 'close') {
    emit(trigger, event)
  }

  if (
    event.defaultPrevented ||
    (trigger === 'backdrop' && noCloseOnBackdropBoolean.value) ||
    (trigger === 'esc' && noCloseOnEscBoolean.value)
  ) {
    emit('hide-prevented')
    return
  }

  emit('hide', event)

  modelValue.value = false
}
```

if I press close button  and I want to prevent modal close (like in test), I emit ```close``` event then close event doesn't  fire ```hide``` if prevented or ```no-close-on-backdrop```/```no-close-on-esc``` are true

the same for BOffcanvas
```js
const buildTriggerableEvent = (
  type: string,
  opts: Partial<BvTriggerableEvent> = {}
): BvTriggerableEvent =>
  new BvTriggerableEvent(type, {
    cancelable: false,
    target: element.value || null,
    relatedTarget: null,
    trigger: null,
    ...opts,
    componentId: computedId.value,
  })

const hide = (trigger = '') => {
  const event = buildTriggerableEvent('hide', {cancelable: trigger !== '', trigger})

  if (trigger === 'close') {
    emit(trigger, event)
  }
  if (trigger === 'esc') {
    emit(trigger, event)
  }

  if (
    event.defaultPrevented ||
    (trigger === 'backdrop' && noCloseOnBackdropBoolean.value) ||
    (trigger === 'esc' && noCloseOnEscBoolean.value)
  ) {
    emit('hide-prevented')
    return
  }

  emit('hide', event)

  modelValue.value = false
}
``` 
I press ESC event ```esc``` should be fired but it doesn't fire ```hide``` if prevented  or ```no-close-on-backdrop```/```no-close-on-esc``` are true

**Comment by VividLemon** [comment]
> I think this is better and modal.spec.ts doesn't fail

Then fix the tests. A failing test is only an issue if the expected behavior is not correct. (Too many double negatives? If something was not correct when the test was written, then it's just a bad test!)

> if I press close button and I want to prevent modal close (like in test), I emit close event then close event doesn't fire hide if prevented or no-close-on-backdrop/no-close-on-esc are true

Let's get two things straightened out here. First off, in my code example you're able to prevent modal closing in all cases. Second off, you're misunderstanding the purpose of the hide event. It is not to indicate that the modal has been hidden, that's what the hidden event is for. Functionally it is in line with https://getbootstrap.com/docs/5.3/components/modal/#events , though our system utilizes a more extra events for specific targets. 

To add to this, the hide event also emits the Event, which allows it to be preventable. Those events are grouped up because they all signify the same point in time for the event... Assuming that makes sense. 

So, my code is in-line with what the spec is for Bootstrap v5 and how I would prefer it to be. Unless you can tell me if and how it is functionally not in compliance with the Bootstrap v5, then that is how I would prefer it to be fixed. 

Though hypothetically the 
```ts
if (
    event.defaultPrevented ||
    (trigger === 'backdrop' && noCloseOnBackdropBoolean.value) ||
    (trigger === 'esc' && noCloseOnEscBoolean.value)
  ) {
    emit('hide-prevented', event)
    return
  }
```
`event.defaultPrevented` for the first check is redundant, as you couldn't have preventedDefault yet. 

Functionally the movement of event emitting like `close` `ok`, etc is minor, but when you put it into the context of it being the same as the hide event, then I think it makes more sense. But you need to know what the purpose of the hide event is for above all else.

File: packages/bootstrap-vue-next/src/components/BOffcanvas/BOffcanvas.vue [hunk 1]
```diff
@@ -284,6 +276,15 @@ const hide = (trigger = '') => {
...
   }
+
+  if (trigger === 'close') {
+    emit(trigger, event)
+  }
+  if (trigger === 'esc') {
+    emit(trigger, event)
+  }
+  emit('hide', event)
+
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/components/BOffcanvas/BOffcanvas.vue:287]:
```suggestion

if(event.defaultPrevented) {
    emit('hide-prevented')
    return
}
```

We need to ensure that you can prevent the closing when this event is triggered

File: packages/bootstrap-vue-next/src/components/BOffcanvas/BOffcanvas.vue [hunk 2]
```diff
@@ -268,14 +268,6 @@ const buildTriggerableEvent = (
...
-  if (trigger === 'close') {
-    emit(trigger, event)
-  }
-  if (trigger === 'esc') {
-    emit(trigger, event)
-  }
-  emit('hide', event)
-
   if (
     event.defaultPrevented ||
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/components/BOffcanvas/BOffcanvas.vue:272]:
```suggestion
```
You can't prevent default at this point, so this check is unnecessary

Comment (VividLemon) [reply packages/bootstrap-vue-next/src/components/BOffcanvas/BOffcanvas.vue:272]: There might be some linting stuff to look at with this change

File: packages/bootstrap-vue-next/src/components/BModal.vue [hunk 3]
```diff
@@ -463,16 +471,13 @@ const hide = (trigger = '') => {
   if (trigger === 'close') {
     emit(trigger, event)
   }
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/components/BModal.vue:473]:
```suggestion
  }
    emit('hide', event)
```

Comment (VividLemon) [reply packages/bootstrap-vue-next/src/components/BModal.vue:473]: I think the suggestion contains a linting tab issue

File: packages/bootstrap-vue-next/src/components/BModal.vue [hunk 4]
```diff
@@ -463,16 +471,13 @@ const hide = (trigger = '') => {
...
-    event.defaultPrevented ||
-    (trigger === 'backdrop' && noCloseOnBackdropBoolean.value) ||
-    (trigger === 'esc' && noCloseOnEscBoolean.value)
-  ) {
+  if (event.defaultPrevented) {
     emit('hide-prevented')
     return
   }
+
+  emit('hide', event)
```
Comment (VividLemon) [inline packages/bootstrap-vue-next/src/components/BModal.vue:479]:
```suggestion
```
This is moved up with the others, (close, ok, etc)

**Review by VividLemon — CHANGES_REQUESTED** [review]

**Comment by makroxyz** [comment]
Hi I commited changes but I'm not sure this is right  in both files
```js
 emit('hide', event)

  if (event.defaultPrevented) {
    emit('hide-prevented')
    return
  }
```
I emit hide before checking if it's prevented...

**Comment by VividLemon** [comment]
> Hi I commited changes but I'm not sure this is right in both files
> 
> ```js
>  emit('hide', event)
> 
>   if (event.defaultPrevented) {
>     emit('hide-prevented')
>     return
>   }
> ```
> 
> I emit hide before checking if it's prevented...

Yes, that's the point. it's so you can do things like 
```html
<BModal @hide.prevent @hide-prevented="consoleLog('foo')" />
```

**Comment by makroxyz** [comment]
@VividLemon 
I'm facing this issues (is it really an issue?)

When a BModal is conditioned by v-if, ```show/shown``` events are not fired. Old Bootstrap-vue does it.
I think because event are fired on BTransition tag.

```js
<template>
  <BContainer>
    <BRow>
      <BCol> 
        <BModal v-if="show" v-model="show" @show="() => console.log('show')">
          test
        </BModal>
        <BButton @click="show = true">Test</BButton>
      </BCol>
    </BRow>
  </BContainer>
</template>

<script setup lang="ts">
import { ref } from 'vue';
const show = ref(false)
</script>
````

**Comment by VividLemon** [comment]
> @VividLemon I'm facing this issue
> 
> When a BModal is conditioned by v-if, `show/shown` events are not fired. Old Bootstrap-vue does it. I think because event are fired on BTransition tag.
> 
> ```js
> <template>
>   <BContainer>
>     <BRow>
>       <BCol> 
>         <BModal v-if="show" v-model="show" @show="() => console.log('show')">
>           test
>         </BModal>
>         <BButton @click="show = true">Test</BButton>
>       </BCol>
>     </BRow>
>   </BContainer>
> </template>
> 
> <script setup lang="ts">
> import { ref } from 'vue';
> const show = ref(false)
> </script>
> ```
> 
> Update: the same when a modal is used in dynamic component
> 
> ```
> <component :is="modalComponent" v-show="show" />   // fires show* event only on SECOND open since modalComponent can be undefined or null
> ```
> 
> https://vuejs.org/guide/built-ins/transition.html#the-transition-component
> 
> dynamic component or v-if component should be inside transition

This would be weird for a person to do. But regardless, I'm not sure how to fix it. Unlike attributes, directives don't fall through to $attrs. So there is no way for the component to know if a directive is attached to it (that I know of)

**Review by VividLemon — CHANGES_REQUESTED** [review]
This is the only remaining feedback I have.

**Comment by makroxyz** [comment]
> > @VividLemon I'm facing this issue
> > When a BModal is conditioned by v-if, `show/shown` events are not fired. Old Bootstrap-vue does it. I think because event are fired on BTransition tag.
> > ```js
> > <template>
> >   <BContainer>
> >     <BRow>
> >       <BCol> 
> >         <BModal v-if="show" v-model="show" @show="() => console.log('show')">
> >           test
> >         </BModal>
> >         <BButton @click="show = true">Test</BButton>
> >       </BCol>
> >     </BRow>
> >   </BContainer>
> > </template>
> > 
> > <script setup lang="ts">
> > import { ref } from 'vue';
> > const show = ref(false)
> > </script>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Update: the same when a modal is used in dynamic component
> > ```
> > <component :is="modalComponent" v-show="show" />   // fires show* event only on SECOND open since modalComponent can be undefined or null
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > https://vuejs.org/guide/built-ins/transition.html#the-transition-component
> > dynamic component or v-if component should be inside transition
> 
> This would be weird for a person to do. But regardless, I'm not sure how to fix it. Unlike attributes, directives don't fall through to $attrs. So there is no way for the component to know if a directive is attached to it (that I know of)

That's right. But this could be a problem... I don't have any idea on how to solve it. While a v-if on a BModal should be "managed" by ```lazy``` attribute, using dynamic component isn't a bad practice and can be an issue

**Comment by VividLemon** [comment]
> That's right. But this could be a problem... I don't have any idea on how to solve it. While a v-if on a BModal should be "managed" by `lazy` attribute, using dynamic component isn't a bad practice and can be an issue

You should make an issue ticket for this

**Pull request merged**